### PR TITLE
Dotty #4013

### PR DESCRIPTION
**Pull request by allanrenucci**
This version of Dotty fixes several unsoundness issues and remove the unsound protected[this] escape hatch for variance checking. See lampepfl/dotty#4013

@julienrf You might want to play with this branch and make it compile with Dotty

### CODE REVIEW

File: collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala [hunk 1]
```diff
@@ -125,8 +125,8 @@ sealed abstract class ImmutableArray[+A]
   * @define coll immutable array
   * @define Coll `ImmutableArray`
   */
-object ImmutableArray extends StrictOptimizedClassTagSeqFactory[ImmutableArray] { self =>
-  val untagged: SeqFactory[ImmutableArray] = new ClassTagSeqFactory.AnySeqDelegate(self)
+object ImmutableArray extends StrictOptimizedClassTagSeqFactory[ImmutableArray] {
+  val untagged: SeqFactory[ImmutableArray] = new ClassTagSeqFactory.AnySeqDelegate(this)
```
Comment (szeiger) [inline collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala]: What's the reason for these changes? Just removing some unnecessary code, or is there a change in Dotty that necessitates this?

Comment (allanrenucci) [reply collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala]: We recently disallowed named self in lampepfl/dotty#4118. We might reconsider it though

Comment (odersky) [reply collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala]: But the fix refers to the wrong this! Btw we will revert #4118.

**Comment by odersky** [comment]
Superseded by #519.
